File: 2021년/7월/0715 이분 그래프.cpp
// 백준_1707_이분 그래프_DFS_골드 4

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
 
int k, v, e, flag, check[20001],visited[20001];
vector<int> graph[20001];

void input_color(int now)
{
	for(int i=0;i<graph[now].size();i++)
	{
		int next = graph[now][i];
		
		if(check[next] == 0)
		{
			if(check[now] == 1) check[next] = 2;
			else check[next] = 1;
			
			input_color(next);
		}
	}
}

void solution(int now)
{
	for(int i=0;i<graph[now].size();i++)
	{
		int next = graph[now][i];
		if(check[next] == check[now]) 
		{
			flag = 1;
			return ;
		}
		
		if(visited[next] == 0)
		{
			visited[next] = 1;
			solution(next);
		}
		
		if (flag == 1) return;
	}
}

int main()
{
	cin >> k;
	while(k--)
	{
		cin >> v >> e;
		
		for(int i=1;i<v+1;i++)
		{
			graph[i].resize(0);
			check[i] = 0;
			visited[i] = 0;
		}
		
		for(int i=0;i<e;i++)
		{
			int x,y;
			cin >> x >> y;
			graph[x].push_back(y);
			graph[y].push_back(x);	
		}
		
		for(int i=1;i<v+1;i++)
		{
			if(check[i] == 0)
			{
				check[i] = 1;
				input_color(i);
			}
		}
		
		flag = 0;
		for(int i=1;i<v+1;i++)
		{
			if(visited[i] == 0)
			{
				visited[i] = 1;
				solution(i);
			}
			
			if(flag == 1) break;
		}
		
		if(flag == 1) cout << "NO" << "\n";
		else cout << "YES" << "\n";
	}	
} 

File: 2021년/7월/0715 이분 그래프.py
# 백준_1707_이분 그래프_DFS_골드 4

import sys
input = sys.stdin.readline


def dfs(now):
    for i in range(len(graph[now])):
        next = graph[now][i]
        if color[next] == 0:
            if color[now] == 1:
                color[next] = 2
            elif color[now] == 2:
                color[next] = 1
            dfs(next)


def solution(now):
    global flag
    for i in range(len(graph[now])):
        next = graph[now][i]

        if color[next] == color[now]:
            flag = 1
            return

        if visited[next] == 0:
            visited[next] = 1
            solution(next)

        if flag:
            return


for _ in range(int(input())):
    v, e = map(int, input().split())
    graph = [[] for _ in range(v+1)]
    color = [0]*(v+1)
    visited = [0]*(v+1)

    for _ in range(e):
        x, y = map(int, input().split())
        graph[x].append(y)
        graph[y].append(x)

    for i in range(1, v+1):
        if color[i] == 0:
            color[i] = 1
            dfs(i)

    flag = 0
    for i in range(1, v+1):
        if visited[i] == 0:
            visited[i] = 1
            solution(i)

        if flag:
            break

    if flag:
        print("NO")
    else:
        print("YES")
